#### Project_Week10_Shenoy-Kamath

# MSDS420 - Project - Group 6 - Retail Customer Behavior For Growth - 

### Approach: Docker with MySQL  - Surrogate Key Vs Composite Key tables - Benchmarking 

In [1]:
pip install -q pymysql pandas

Note: you may need to restart the kernel to use updated packages.


In [1]:
import pymysql, time, pandas as pd
import os

In [2]:
# --- clean connect to MySQL on 127.0.0.1:3307 using .env ---

# 1) Load env (override any cached values)
from dotenv import load_dotenv
import os, socket
from sqlalchemy import create_engine, text
import pandas as pd

load_dotenv(".env", override=True)

host = os.getenv("MYSQL_HOST")
port = os.getenv("MYSQL_PORT")
user = os.getenv("MYSQL_USER")
pwd  = os.getenv("MYSQL_PASSWORD")
db   = os.getenv("MYSQL_DB")

print("Loaded from .env ->", {"host": host, "port": port, "db": db, "user": user})

# 2) Assert the values we NEED
if host != "127.0.0.1":
    raise ValueError(f"MYSQL_HOST must be 127.0.0.1, got {host!r}")
if str(port) != "3307":
    raise ValueError(f"MYSQL_PORT must be 3307 (per docker ps), got {port!r}")
if not all([user, pwd, db]):
    raise ValueError("Missing MYSQL_USER / MYSQL_PASSWORD / MYSQL_DB")

# 3) Prove the TCP port is listening
sock = socket.create_connection((host, int(port)), timeout=5)
sock.close()
print("Socket check OK -> can reach", f"{host}:{port}")

# 4) Build engine with these exact values (no reuse)
url = f"mysql+pymysql://{user}:{pwd}@{host}:{port}/{db}"
print("Connecting:", url.replace(pwd, "******"))
engine = create_engine(url, pool_pre_ping=True, connect_args={"connect_timeout": 5})

# 5) Smoke test
def q(sql, **params):
    return pd.read_sql_query(text(sql), engine, params=params)

print(q("SELECT VERSION() AS mysql_version, DATABASE() AS current_db;"))

Loaded from .env -> {'host': '127.0.0.1', 'port': '3307', 'db': 'retail_db', 'user': 'root'}
Socket check OK -> can reach 127.0.0.1:3307
Connecting: mysql+pymysql://root:******@127.0.0.1:3307/retail_db
  mysql_version current_db
0        8.0.43  retail_db


In [3]:
from dotenv import load_dotenv
import os

# Load .env from current folder
load_dotenv(dotenv_path=".env")

# Print values to confirm (masking password)
print("MYSQL_USER:", os.getenv("MYSQL_USER"))
print("MYSQL_HOST:", os.getenv("MYSQL_HOST"))
print("MYSQL_PORT:", os.getenv("MYSQL_PORT"))
print("MYSQL_DB:", os.getenv("MYSQL_DB"))

MYSQL_USER: root
MYSQL_HOST: 127.0.0.1
MYSQL_PORT: 3307
MYSQL_DB: retail_db


In [4]:
# render high resolution plots
%config InlineBackend.figure_format = "retina"

In [5]:
# helper function to take an SQL string and optional parameters
from sqlalchemy import text
import pandas as pd

# Simple helper: run SQL and return a pandas DataFrame
def q(sql, **params):
    return pd.read_sql_query(text(sql), engine, params=params)

In [6]:
# smoke test
q("SELECT DATABASE() AS db;")

db
0  retail_db

# PART 1 - Performance for Assignment specific queries

In [7]:
# helper snippets (inline so no session vars needed)
STORE   = "(SELECT store_sk FROM bench_lines_data LIMIT 1)"
DAY_LO  = "(SELECT MIN(`day`) FROM bench_lines_data)"
DAY_HI  = f"({DAY_LO} + 30)"  # ~30-day window
HH      = "(SELECT household_sk FROM bench_lines_data LIMIT 1)" 

In [9]:
# ---- connection ----
CONN = dict(host="127.0.0.1", port=3307, user="bench", password="benchpw", database="retail_db")

def run(sql):
    with pymysql.connect(**CONN, cursorclass=pymysql.cursors.DictCursor, autocommit=True) as conn:
        with conn.cursor() as cur:
            t0 = time.perf_counter()
            cur.execute(sql)
            rows = cur.fetchall()
            ms = (time.perf_counter() - t0) * 1000
            return rows, ms

def choose_store_and_window():
    """Pick a busy store and a 30-day window that actually has rows."""
    with pymysql.connect(**CONN, cursorclass=pymysql.cursors.DictCursor, autocommit=True) as conn:
        with conn.cursor() as cur:
            # 1) pick the busiest store in CK (SK mirrors data)
            cur.execute("""
                SELECT store_sk, MIN(`day`) AS min_day, MAX(`day`) AS max_day, COUNT(*) AS cnt
                FROM bench_lines_ck
                GROUP BY store_sk
                ORDER BY cnt DESC
                LIMIT 1
            """)
            rec = cur.fetchone()
            store = int(rec["store_sk"])
            min_day, max_day = int(rec["min_day"]), int(rec["max_day"])

            # first attempt: earliest 31-day window
            day_lo = min_day
            day_hi = min(min_day + 30, max_day)

            cur.execute("""
                SELECT COUNT(*) AS n
                FROM bench_lines_ck
                WHERE store_sk=%s AND `day` BETWEEN %s AND %s
            """, (store, day_lo, day_hi))
            n = cur.fetchone()["n"]

            # fallback: center on the densest single day, widen ±7
            if n == 0:
                cur.execute("""
                    SELECT `day`, COUNT(*) AS c
                    FROM bench_lines_ck
                    WHERE store_sk=%s
                    GROUP BY `day`
                    ORDER BY c DESC
                    LIMIT 1
                """, (store,))
                best_day = int(cur.fetchone()["day"])
                day_lo = max(min_day, best_day - 7)
                day_hi = min(max_day, best_day + 7)

            return store, day_lo, day_hi

# ---- choose constants for this run ----
STORE, DAY_LO, DAY_HI = choose_store_and_window()
print(f"Benchmark window: store_sk={STORE}, day {DAY_LO}..{DAY_HI}")

# ---- build queries USING the constants above ----
QUERIES = {
    # 1) Track customer spending trends (by product within store & 30-day window)
    "spend_trends": {
        "ck": f"""
            SELECT product_sk, SUM(sales) AS revenue
            FROM bench_lines_ck
            WHERE store_sk = {STORE} AND `day` BETWEEN {DAY_LO} AND {DAY_HI}
            GROUP BY product_sk
        """,
        "sk": f"""
            SELECT product_sk, SUM(sales) AS revenue
            FROM bench_lines_sk
            WHERE store_sk = {STORE} AND `day` BETWEEN {DAY_LO} AND {DAY_HI}
            GROUP BY product_sk
        """
    },

    # 2) Evaluate demographic influences on customer spend
    "demo_influence": {
        "ck": f"""
            SELECT h.income_desc, h.age_desc, SUM(b.sales) AS revenue
            FROM bench_lines_ck b
            JOIN household_dim h ON h.household_sk = b.household_sk
            WHERE b.`day` BETWEEN {DAY_LO} AND {DAY_HI}
            GROUP BY h.income_desc, h.age_desc
            ORDER BY revenue DESC
            LIMIT 10
        """,
        "sk": f"""
            SELECT h.income_desc, h.age_desc, SUM(b.sales) AS revenue
            FROM bench_lines_sk b
            JOIN household_dim h ON h.household_sk = b.household_sk
            WHERE b.`day` BETWEEN {DAY_LO} AND {DAY_HI}
            GROUP BY h.income_desc, h.age_desc
            ORDER BY revenue DESC
            LIMIT 10
        """
    },

    # 3) Identify growth categories (current 30d vs previous 28d) using product_dim.commodity_desc
    "growth_categories": {
        "ck": f"""
            WITH cur AS (
              SELECT p.commodity_desc AS category, SUM(b.sales) AS rev
              FROM bench_lines_ck b
              JOIN product_dim p ON p.product_sk = b.product_sk
              WHERE b.`day` BETWEEN {DAY_LO} AND {DAY_HI}
              GROUP BY p.commodity_desc
            ),
            prev AS (
              SELECT p.commodity_desc AS category, SUM(b.sales) AS rev
              FROM bench_lines_ck b
              JOIN product_dim p ON p.product_sk = b.product_sk
              WHERE b.`day` BETWEEN ({DAY_LO} - 28) AND ({DAY_LO} - 1)
              GROUP BY p.commodity_desc
            )
            SELECT c.category,
                   c.rev AS rev_cur,
                   COALESCE(p.rev,0) AS rev_prev,
                   CASE WHEN COALESCE(p.rev,0)=0 THEN NULL
                        ELSE (c.rev - p.rev)/p.rev END AS growth_ratio
            FROM cur c
            LEFT JOIN prev p USING (category)
            ORDER BY growth_ratio DESC
            LIMIT 10
        """,
        "sk": f"""
            WITH cur AS (
              SELECT p.commodity_desc AS category, SUM(b.sales) AS rev
              FROM bench_lines_sk b
              JOIN product_dim p ON p.product_sk = b.product_sk
              WHERE b.`day` BETWEEN {DAY_LO} AND {DAY_HI}
              GROUP BY p.commodity_desc
            ),
            prev AS (
              SELECT p.commodity_desc AS category, SUM(b.sales) AS rev
              FROM bench_lines_sk b
              JOIN product_dim p ON p.product_sk = b.product_sk
              WHERE b.`day` BETWEEN ({DAY_LO} - 28) AND ({DAY_LO} - 1)
              GROUP BY p.commodity_desc
            )
            SELECT c.category,
                   c.rev AS rev_cur,
                   COALESCE(p.rev,0) AS rev_prev,
                   CASE WHEN COALESCE(p.rev,0)=0 THEN NULL
                        ELSE (c.rev - p.rev)/p.rev END AS growth_ratio
            FROM cur c
            LEFT JOIN prev p USING (category)
            ORDER BY growth_ratio DESC
            LIMIT 10
        """
    },

    # 4) Measure the impact of marketing campaigns
    "campaign_impact": {
        "ck": """
            WITH redeemed AS (
              SELECT DISTINCT tf.product_sk, tf.store_sk, tf.`day`
              FROM coupon_redemption_fact crf
              JOIN bench_lines_ck tf
                ON tf.household_sk = crf.household_sk
               AND tf.product_sk   = crf.product_sk
               AND tf.`day`        = crf.`day`
            )
            SELECT 'redeemed' AS flag, SUM(b.sales) AS revenue, SUM(b.qty) AS units
            FROM bench_lines_ck b
            JOIN redeemed r USING (product_sk, store_sk, `day`)
            UNION ALL
            SELECT 'not_redeemed', SUM(b.sales), SUM(b.qty)
            FROM bench_lines_ck b
            LEFT JOIN redeemed r USING (product_sk, store_sk, `day`)
            WHERE r.product_sk IS NULL
        """,
        "sk": """
            WITH redeemed AS (
              SELECT DISTINCT tf.product_sk, tf.store_sk, tf.`day`
              FROM coupon_redemption_fact crf
              JOIN bench_lines_sk tf
                ON tf.household_sk = crf.household_sk
               AND tf.product_sk   = crf.product_sk
               AND tf.`day`        = crf.`day`
            )
            SELECT 'redeemed' AS flag, SUM(b.sales) AS revenue, SUM(b.qty) AS units
            FROM bench_lines_sk b
            JOIN redeemed r USING (product_sk, store_sk, `day`)
            UNION ALL
            SELECT 'not_redeemed', SUM(b.sales), SUM(b.qty)
            FROM bench_lines_sk b
            LEFT JOIN redeemed r USING (product_sk, store_sk, `day`)
            WHERE r.product_sk IS NULL
        """
    }
}

def main():
    results = []
    for task, pair in QUERIES.items():
        for design in ("ck", "sk"):
            rows, ms = run(pair[design])
            results.append({
                "assignment_task": task,
                "design": design,
                "ms": round(ms, 2),
                "rowcount": len(rows)
            })
    df = pd.DataFrame(results).sort_values(["assignment_task", "design"])
    print(df.to_string(index=False))
    df.to_csv("benchmark_results.csv", index=False)
    print("\nResults saved to benchmark_results.csv")

main()


Benchmark window: store_sk=140, day 6..36
  assignment_task design      ms  rowcount
  campaign_impact     ck 6744.19         2
  campaign_impact     sk 6119.41         2
   demo_influence     ck  563.75        10
   demo_influence     sk  488.05        10
growth_categories     ck 1561.09        10
growth_categories     sk 1247.66        10
     spend_trends     ck    5.47       177
     spend_trends     sk    6.63       177

Results saved to benchmark_results.csv


## Assignment Benchmarks — Store 140, Days 6..36

### Results (ms)
**Benchmark window:** `store_sk = 140`, `days 6..36`

| assignment_task   | design |    ms   | rowcount |
|-------------------|:------:|--------:|---------:|
| campaign_impact   |  ck    | 6744.19 |        2 |
| campaign_impact   |  sk    | 6119.41 |        2 |
| demo_influence    |  ck    |  563.75 |       10 |
| demo_influence    |  sk    |  488.05 |       10 |
| growth_categories |  ck    | 1561.09 |       10 |
| growth_categories |  sk    | 1247.66 |       10 |
| spend_trends      |  ck    |    5.47 |      177 |
| spend_trends      |  sk    |    6.63 |      177 |


### Quick takeaways
- **SK is faster on 3 of 4 tasks; CK wins `spend_trends` in this run.**  
  - *spend_trends:* **CK ≈ 17% faster** (5.47 vs 6.63 ms) — CK’s clustering on `(store, day, product, …)` gives great locality.  
  - *demo_influence:* **SK ≈ 13% faster** (488.05 vs 563.75 ms) — both scan/aggregate; SK edges out.  
  - *growth_categories:* **SK ≈ 20% faster** (1247.66 vs 1561.09 ms) — join + group; SK benefits from secondary index paths.  
  - *campaign_impact:* **SK ≈ 9% faster** (6119.41 vs 6744.19 ms) — heavy join/aggregate; differences are modest.

### Interpretation
- For **aligned access (store + day)**, both designs are efficient; in this sample **CK outperformed SK** on `spend_trends` due to clustered PK order.
- For **join-heavy aggregations** (demo/growth/campaign), the fact-table scan + grouping dominates, so differences remain moderate; **SK often has a slight edge** thanks to targeted secondary indexes.
- Remember the microbenchmarks:
  - **Household-centric lookups:** **SK wins big** via `idx_household` (CK must scan).
  - **Insert/maintenance & storage:** **CK wins** — smaller tables, fewer/lighter indexes → faster bulk inserts and table maintenance.


# Part 2 - Comparsion between Surrogate key and Composite key

## 1) Table & Index size (space overhead)

In [10]:
# Table + index bytes

def run_sql(query):
    with pymysql.connect(**CONN, cursorclass=pymysql.cursors.DictCursor, autocommit=True) as conn:
        with conn.cursor() as cur:
            cur.execute(query)
            return cur.fetchall()

# ---- 1) Table + index bytes (MB) ----
q_table_sizes = """
SELECT table_name, 
       ROUND(data_length/1024/1024,2)  AS data_mb,
       ROUND(index_length/1024/1024,2) AS index_mb,
       ROUND((data_length+index_length)/1024/1024,2) AS total_mb
FROM information_schema.tables
WHERE table_schema='retail_db'
  AND table_name IN ('bench_lines_ck','bench_lines_sk')
ORDER BY table_name;
"""

table_sizes_df = pd.DataFrame(run_sql(q_table_sizes))
print("Table sizes (MB):")
print(table_sizes_df.to_string(index=False))
table_sizes_df.to_csv("table_sizes.csv", index=False)

# ---- 2) Per-index details 
q_index_details = """
SELECT 
    table_name, 
    index_name, 
    NON_UNIQUE, 
    SEQ_IN_INDEX, 
    COLUMN_NAME, 
    SUB_PART, 
    INDEX_TYPE,
    CARDINALITY
FROM information_schema.STATISTICS
WHERE table_schema='retail_db'
  AND table_name IN ('bench_lines_ck','bench_lines_sk')
ORDER BY table_name, index_name, SEQ_IN_INDEX;
"""

index_details_df = pd.DataFrame(run_sql(q_index_details))
print("\nIndex details:")
print(index_details_df.to_string(index=False))
index_details_df.to_csv("index_details.csv", index=False)

print("\n Saved: table_sizes.csv, index_details.csv")

Table sizes (MB):
    TABLE_NAME data_mb index_mb total_mb
bench_lines_ck   98.69     0.00    98.69
bench_lines_sk  109.66    75.17   184.83

Index details:
    TABLE_NAME         INDEX_NAME  NON_UNIQUE  SEQ_IN_INDEX  COLUMN_NAME SUB_PART INDEX_TYPE  CARDINALITY
bench_lines_ck            PRIMARY           0             1     store_sk     None      BTREE          407
bench_lines_ck            PRIMARY           0             2          day     None      BTREE        52770
bench_lines_ck            PRIMARY           0             3   product_sk     None      BTREE      1410659
bench_lines_ck            PRIMARY           0             4 household_sk     None      BTREE      1417842
bench_lines_sk      idx_household           1             1 household_sk     None      BTREE          796
bench_lines_sk idx_store_day_prod           1             1     store_sk     None      BTREE          393
bench_lines_sk idx_store_day_prod           1             2          day     None      BTREE        5

##  Table & Index Sizes Benchmarks

### Table Sizes
| Table            | Data (MB) | Index (MB) | Total (MB) |
|------------------|-----------|------------|------------|
| `bench_lines_ck` | 98.69     | 0.00       | 98.69      |
| `bench_lines_sk` | 109.66    | 75.17      | 184.83     |

- **Composite Key (CK)** table is smaller overall because it only has a clustered **PRIMARY KEY** on `(store_sk, day, product_sk, household_sk)`.  
- **Surrogate Key (SK)** table consumes almost **2× the space** because:
  - The `AUTO_INCREMENT id` primary key is stored in **every secondary index**.
  - Additional indexes (`idx_household`, `idx_store_day_prod`) were added to support diverse queries.  

---

### Index Details
- **CK Table (`bench_lines_ck`):**
  - One PRIMARY KEY spanning **4 columns** (`store_sk → day → product_sk → household_sk`).
  - No secondary indexes defined.
  - Efficient for queries aligned with this leftmost prefix (e.g., “sales by store/date window”).
  - Not efficient for household-centric lookups without full or wide scans.

- **SK Table (`bench_lines_sk`):**
  - PRIMARY KEY only on `id` (compact BIGINT).
  - Has secondary indexes:
    - `idx_household` on `household_sk`
    - `idx_store_day_prod` on `(store_sk, day, product_sk)`
  - Provides **flexible access paths** for varied queries (household-centric, store/day ranges).
  - Larger storage overhead because every secondary index entry carries the `id` PK.

---

## Conclusions

1. **Storage Tradeoff**  
   - CK schema is **leaner** (~99 MB total).  
   - SK schema is **heavier** (~185 MB total) due to PK + multiple secondary indexes.  

2. **Performance Implications**  
   - CK performs best for queries that match its **natural clustering order** (store → day → product → household).  
   - SK supports **diverse query patterns** because you can add targeted indexes, at the cost of storage.  

3. **Maintainability**  
   - SK design is **more flexible**


##### NOTE:  for the INSERT throughput I had to do this first to grant permissions : 

docker exec -it sk_mysql mysql -uroot -p retail_db

--sql
GRANT CREATE, DROP, INSERT ON retail_db.* TO 'bench'@'%';
FLUSH PRIVILEGES;


## 2) Insert throughput (append vs. scattered)

In [11]:
import os
# ---- run insert-throughput inside ONE session so @t0/@t1/@t2 persist ----
with pymysql.connect(**CONN, cursorclass=pymysql.cursors.DictCursor, autocommit=True) as conn:
    with conn.cursor() as cur:
        # fresh clones
        cur.execute("DROP TABLE IF EXISTS bench_ins_ck")
        cur.execute("DROP TABLE IF EXISTS bench_ins_sk")
        cur.execute("CREATE TABLE bench_ins_ck LIKE bench_lines_ck")
        cur.execute("CREATE TABLE bench_ins_sk LIKE bench_lines_sk")

        # time CK insert
        cur.execute("SET @t0 := NOW(6)")
        cur.execute("INSERT INTO bench_ins_ck SELECT * FROM bench_lines_data")
        cur.execute("SET @t1 := NOW(6)")

        # time SK insert
        cur.execute("""
            INSERT INTO bench_ins_sk (store_sk, `day`, product_sk, household_sk, sales, qty)
            SELECT store_sk, `day`, product_sk, household_sk, sales, qty
            FROM bench_lines_data
        """)
        cur.execute("SET @t2 := NOW(6)")

        # fetch timings
        cur.execute("""
            SELECT 
              TIMESTAMPDIFF(MICROSECOND,@t0,@t1)/1000.0 AS ms_insert_ck,
              TIMESTAMPDIFF(MICROSECOND,@t1,@t2)/1000.0 AS ms_insert_sk
        """)
        timings = cur.fetchone()

        # rowcounts (nice to include)
        cur.execute("SELECT COUNT(*) AS n FROM bench_ins_ck")
        n_ck = cur.fetchone()["n"]
        cur.execute("SELECT COUNT(*) AS n FROM bench_ins_sk")
        n_sk = cur.fetchone()["n"]

# build two rows to append
rows = [
    {"assignment_task": "insert_throughput", "design": "ck", "ms": round(timings["ms_insert_ck"], 2), "rowcount": n_ck},
    {"assignment_task": "insert_throughput", "design": "sk", "ms": round(timings["ms_insert_sk"], 2), "rowcount": n_sk},
]
new_df = pd.DataFrame(rows)

# append to existing benchmark_results.csv if present, else create
out_path = "benchmark_results.csv"
if os.path.exists(out_path):
    base = pd.read_csv(out_path)
    combined = pd.concat([base, new_df], ignore_index=True)
else:
    combined = new_df

# save combined file (overwrite the same CSV for simplicity)
combined.to_csv(out_path, index=False)

print(new_df.to_string(index=False))
print(f"\n Appended insert timings to {out_path}")


  assignment_task design       ms  rowcount
insert_throughput     ck 23433.62   1423952
insert_throughput     sk 37728.66   1423952

 Appended insert timings to benchmark_results.csv


## Insert Throughput Benchmark

### Results
| Design | Insert Time (ms) | Rows Inserted |
|--------|------------------:|--------------:|
| CK     | 23,433.62         | 1,423,952     |
| SK     | 37,728.66         | 1,423,952     |

### Observations
- **Composite Key (CK)** completed the bulk load in **~23.4 s**, while **Surrogate Key (SK)** took **~37.7 s** — SK is ~**61% slower**, i.e., CK is ~**38% faster** for this load.
- Why:
  - CK’s clustering on `(store_sk, day, product_sk, household_sk)` fits the input order and updates only the clustered index.
  - SK uses an `AUTO_INCREMENT id` **plus two secondary indexes** (`idx_household`, `idx_store_day_prod`), so each insert also updates extra index structures.
- **Rowcount matches** for both (1,423,952), confirming correctness.

### Conclusion
- **CK** is more efficient for large batch loads due to fewer index updates.
- **SK** provides flexible query access paths but at a higher **write/maintenance cost** during bulk inserts.

## 3) Random point lookups (household-centric)

In [12]:
def run_explain(sql):
    with pymysql.connect(**CONN, cursorclass=pymysql.cursors.DictCursor, autocommit=True) as conn:
        with conn.cursor() as cur:
            cur.execute(sql)
            return [row for row in cur.fetchall()]

# 1) Pick a household_sk to test
with pymysql.connect(**CONN, cursorclass=pymysql.cursors.DictCursor, autocommit=True) as conn:
    with conn.cursor() as cur:
        cur.execute("SELECT household_sk FROM bench_lines_data LIMIT 1;")
        hh = cur.fetchone()["household_sk"]

print(f"Testing with household_sk = {hh}")

# 2) Run EXPLAIN ANALYZE for CK and SK
sql_ck = f"EXPLAIN ANALYZE SELECT SUM(qty) AS total_qty FROM bench_lines_ck WHERE household_sk={hh};"
sql_sk = f"EXPLAIN ANALYZE SELECT SUM(qty) AS total_qty FROM bench_lines_sk WHERE household_sk={hh};"

plan_ck = run_explain(sql_ck)
plan_sk = run_explain(sql_sk)

# 3) Pretty print (MySQL returns the execution plan as rows with 'EXPLAIN' key)
print("\n--- Composite Key (CK) ---")
for row in plan_ck:
    print(row["EXPLAIN"])

print("\n--- Surrogate Key (SK) ---")
for row in plan_sk:
    print(row["EXPLAIN"])

Testing with household_sk = 392

--- Composite Key (CK) ---
-> Aggregate: sum(bench_lines_ck.qty)  (cost=163175 rows=1) (actual time=1444..1444 rows=1 loops=1)
    -> Filter: (bench_lines_ck.household_sk = 392)  (cost=148996 rows=141784) (actual time=1.43..1444 rows=1170 loops=1)
        -> Table scan on bench_lines_ck  (cost=148996 rows=1.42e+6) (actual time=1.42..1169 rows=1.42e+6 loops=1)


--- Surrogate Key (SK) ---
-> Aggregate: sum(bench_lines_sk.qty)  (cost=1404 rows=1) (actual time=7.79..7.79 rows=1 loops=1)
    -> Index lookup on bench_lines_sk using idx_household (household_sk=392)  (cost=1287 rows=1170) (actual time=0.764..7.26 rows=1170 loops=1)



## Household-Centric Lookup Benchmark

### Results
| Design | Execution Time (ms) | Rows Examined | Matching Rows | Result Rows |
|--------|--------------------:|--------------:|--------------:|------------:|
| CK     | ~1,444              | ~1,420,000    | 1,170         | 1 (SUM)     |
| SK     | ~7.8                | ~1,170        | 1,170         | 1 (SUM)     |

### Observations
- **Composite Key (CK):**
  - Performs a **full table scan** (~1.42M rows) because `household_sk` is the last column of the PK
    `(store_sk, day, product_sk, household_sk)` and can’t be used via the leftmost-prefix rule.
  - The filter then keeps ~1,170 rows, which are aggregated to a single SUM row.
  - Total time ≈ **1,444 ms**.

- **Surrogate Key (SK):**
  - Uses the secondary index **`idx_household`** to jump straight to the ~1,170 matching rows.
  - Examines only those index entries and aggregates to a single SUM row.
  - Total time ≈ **7.8 ms** — **~185× faster** than CK in this test.

### Conclusion
- CK is **inefficient** for predicates that don’t align to the PK’s leftmost columns (here, `household_sk`).
- SK is **highly flexible**: secondary indexes like `idx_household` provide targeted lookups with minimal IO.
- This benchmark highlights the trade-off: **CK = storage/insert efficiency**, **SK = query flexibility & speed** on non-aligned predicates.

## 4) Leftmost-prefix range scans (CK sweet spot)

In [13]:
def run_explain(sql):
    with pymysql.connect(**CONN, cursorclass=pymysql.cursors.DictCursor, autocommit=True) as conn:
        with conn.cursor() as cur:
            cur.execute(sql)
            return [row["EXPLAIN"] for row in cur.fetchall()]

# 1) pick a store and day window
with pymysql.connect(**CONN, cursorclass=pymysql.cursors.DictCursor, autocommit=True) as conn:
    with conn.cursor() as cur:
        cur.execute("SELECT store_sk FROM bench_lines_data LIMIT 1;")
        store = cur.fetchone()["store_sk"]
        cur.execute("SELECT MIN(`day`) AS d0 FROM bench_lines_data;")
        d0 = cur.fetchone()["d0"]
        d1 = d0 + 30

print(f"Testing with store_sk={store}, day range={d0}..{d1}")

# 2) build queries
sql_ck = f"""
EXPLAIN ANALYZE
SELECT product_sk, SUM(sales) AS revenue
FROM bench_lines_ck
WHERE store_sk={store} AND `day` BETWEEN {d0} AND {d1}
GROUP BY product_sk;
"""

sql_sk = f"""
EXPLAIN ANALYZE
SELECT product_sk, SUM(sales) AS revenue
FROM bench_lines_sk
WHERE store_sk={store} AND `day` BETWEEN {d0} AND {d1}
GROUP BY product_sk;
"""

# 3) run and print results
plan_ck = run_explain(sql_ck)
plan_sk = run_explain(sql_sk)

print("\n--- Composite Key (CK) ---")
for line in plan_ck:
    print(line)

print("\n--- Surrogate Key (SK) ---")
for line in plan_sk:
    print(line)

Testing with store_sk=1, day range=1..31

--- Composite Key (CK) ---
-> Table scan on <temporary>  (actual time=5.54..5.54 rows=0 loops=1)
    -> Aggregate using temporary table  (actual time=5.54..5.54 rows=0 loops=1)
        -> Filter: ((bench_lines_ck.store_sk = 1) and (bench_lines_ck.`day` between 1 and 31))  (cost=1.21 rows=1) (actual time=2.19..2.19 rows=0 loops=1)
            -> Index range scan on bench_lines_ck using PRIMARY over (store_sk = 1 AND 1 <= day <= 31)  (cost=1.21 rows=1) (actual time=0.0614..0.0614 rows=0 loops=1)


--- Surrogate Key (SK) ---
-> Table scan on <temporary>  (actual time=0.0432..0.0432 rows=0 loops=1)
    -> Aggregate using temporary table  (actual time=0.0415..0.0415 rows=0 loops=1)
        -> Index range scan on bench_lines_sk using idx_store_day_prod over (store_sk = 1 AND 1 <= day <= 31), with index condition: ((bench_lines_sk.store_sk = 1) and (bench_lines_sk.`day` between 1 and 31))  (cost=2.21 rows=1) (actual time=0.0291..0.0291 rows=0 loops=1)

## Store + Day Range Lookup Benchmark

**Test:** `store_sk = 1`, `day 1..31`

### Results
| Design | Execution Time | Access Path |
|--------|----------------|-------------|
| CK     | **~5.54 s** (range scan node ~**0.061 s**) | Clustered **PRIMARY** range scan on `(store_sk, day, product_sk, household_sk)` |
| SK     | **~0.043 s** (range scan node ~**0.029 s**) | Secondary index **`idx_store_day_prod`** on `(store_sk, day, product_sk)` |

### Observations
- **Both** plans follow an aligned `(store, day)` access path.  
- In this run, **SK is much faster overall** (≈**5.54 s** vs **0.043 s** total). Its narrower secondary index returns the needed keys quickly, and the top-level operators finish almost immediately.  
- CK does use the clustered PK efficiently (its range scan node is ~61 ms), but the overall query shows **~5.54 s** at the top node, likely reflecting temp/aggregation and/or cache warm-up effects during this sample. Re-running after the buffer pool warms often reduces the CK total substantially.

### Conclusion
- For **aligned (store + day)** predicates, both designs are capable of efficient index range scans; in this sample the **SK plan won decisively**.  
- When reporting, consider both the **range-scan node times** (which were close: 61 ms CK vs 29 ms SK) and the **top-level time** (which can be sensitive to cache state).

## 5) Join fanout cost (dim joins)

In [14]:
def run_explain(sql):
    with pymysql.connect(**CONN, cursorclass=pymysql.cursors.DictCursor, autocommit=True) as conn:
        with conn.cursor() as cur:
            cur.execute(sql)
            return [row["EXPLAIN"] for row in cur.fetchall()]

# CK query
sql_ck = """
EXPLAIN ANALYZE
SELECT h.income_desc, h.age_desc, SUM(b.sales) AS revenue
FROM bench_lines_ck b
JOIN household_dim h ON h.household_sk = b.household_sk
GROUP BY h.income_desc, h.age_desc
ORDER BY revenue DESC
LIMIT 10;
"""

# SK query
sql_sk = """
EXPLAIN ANALYZE
SELECT h.income_desc, h.age_desc, SUM(b.sales) AS revenue
FROM bench_lines_sk b
JOIN household_dim h ON h.household_sk = b.household_sk
GROUP BY h.income_desc, h.age_desc
ORDER BY revenue DESC
LIMIT 10;
"""

# Run and print results
plan_ck = run_explain(sql_ck)
plan_sk = run_explain(sql_sk)

print("\n--- Composite Key (CK) ---")
for line in plan_ck:
    print(line)

print("\n--- Surrogate Key (SK) ---")
for line in plan_sk:
    print(line)


--- Composite Key (CK) ---
-> Limit: 10 row(s)  (actual time=16636..16636 rows=10 loops=1)
    -> Sort: revenue DESC, limit input to 10 row(s) per chunk  (actual time=16636..16636 rows=10 loops=1)
        -> Table scan on <temporary>  (actual time=16635..16635 rows=62 loops=1)
            -> Aggregate using temporary table  (actual time=16635..16635 rows=62 loops=1)
                -> Nested loop inner join  (cost=1.71e+6 rows=1.42e+6) (actual time=2.42..3512 rows=1.42e+6 loops=1)
                    -> Table scan on b  (cost=148996 rows=1.42e+6) (actual time=2.35..1119 rows=1.42e+6 loops=1)
                    -> Single-row index lookup on h using PRIMARY (household_sk=b.household_sk)  (cost=1 rows=1) (actual time=0.00133..0.00139 rows=1 loops=1.42e+6)


--- Surrogate Key (SK) ---
-> Limit: 10 row(s)  (actual time=14531..14531 rows=10 loops=1)
    -> Sort: revenue DESC, limit input to 10 row(s) per chunk  (actual time=14531..14531 rows=10 loops=1)
        -> Table scan on <temporary>

## Demographic Influence Benchmark

### Results
| Design | Execution Time | Access Path |
|--------|----------------|-------------|
| CK     | ~16,636 ms     | Full table scan of fact table (`bench_lines_ck`), nested loop join to `household_dim` (PK lookup), aggregate + sort, LIMIT 10 |
| SK     | ~14,531 ms     | Full table scan of fact table (`bench_lines_sk`), nested loop join to `household_dim` (PK lookup), aggregate + sort, LIMIT 10 |

### Observations
- Both plans **scan ~1.42M fact rows**, join each to `household_dim` via its **PRIMARY KEY**, then **aggregate and sort**.
- **SK is ~13% faster** in this run (14.53s vs 16.64s). The difference is modest and likely reflects cache/warm-up and row layout effects; structurally the plans are the same.
- Nested loop lookups into `household_dim` are cheap in both (single-row PK lookups).

### Conclusion
- This workload is **scan + group dominated**, so neither design has a strong structural advantage.
- Expect small run-to-run variation; with a warm buffer pool the gap may narrow.


## 6) Covering-index vs. back-to-table lookups

In [15]:
def run_explain(sql):
    with pymysql.connect(**CONN, cursorclass=pymysql.cursors.DictCursor, autocommit=True) as conn:
        with conn.cursor() as cur:
            cur.execute(sql)
            return [row["EXPLAIN"] for row in cur.fetchall()]

# 1) Add covering indexes (safe to re-run, MySQL will error if already exists)
with pymysql.connect(**CONN, cursorclass=pymysql.cursors.DictCursor, autocommit=True) as conn:
    with conn.cursor() as cur:
        try:
            cur.execute("ALTER TABLE bench_lines_sk ADD KEY cov_store_day_prod_sales (store_sk, `day`, product_sk, sales)")
        except Exception as e:
            print("SK index already exists (or error):", e)
        try:
            cur.execute("ALTER TABLE bench_lines_ck ADD KEY cov_store_day_prod_sales (store_sk, `day`, product_sk, sales)")
        except Exception as e:
            print("CK index already exists (or error):", e)

# 2) Pick a store and date window
with pymysql.connect(**CONN, cursorclass=pymysql.cursors.DictCursor, autocommit=True) as conn:
    with conn.cursor() as cur:
        cur.execute("SELECT store_sk FROM bench_lines_data LIMIT 1;")
        store = cur.fetchone()["store_sk"]
        cur.execute("SELECT MIN(`day`) AS d0 FROM bench_lines_data;")
        d0 = cur.fetchone()["d0"]
        d1 = d0 + 30

print(f"Testing with store_sk={store}, day range={d0}..{d1}")

# 3) Build queries
sql_sk = f"""
EXPLAIN ANALYZE
SELECT product_sk, SUM(sales) AS revenue
FROM bench_lines_sk
WHERE store_sk={store} AND `day` BETWEEN {d0} AND {d1}
GROUP BY product_sk;
"""

sql_ck = f"""
EXPLAIN ANALYZE
SELECT product_sk, SUM(sales) AS revenue
FROM bench_lines_ck
WHERE store_sk={store} AND `day` BETWEEN {d0} AND {d1}
GROUP BY product_sk;
"""

# 4) Run and print results
plan_sk = run_explain(sql_sk)
plan_ck = run_explain(sql_ck)

print("\n--- Surrogate Key (SK) with covering index ---")
for line in plan_sk:
    print(line)

print("\n--- Composite Key (CK) with covering index ---")
for line in plan_ck:
    print(line)

SK index already exists (or error): (1142, "ALTER command denied to user 'bench'@'172.24.0.1' for table 'bench_lines_sk'")
CK index already exists (or error): (1142, "ALTER command denied to user 'bench'@'172.24.0.1' for table 'bench_lines_ck'")
Testing with store_sk=1, day range=1..31

--- Surrogate Key (SK) with covering index ---
-> Table scan on <temporary>  (actual time=0.0257..0.0257 rows=0 loops=1)
    -> Aggregate using temporary table  (actual time=0.0243..0.0243 rows=0 loops=1)
        -> Index range scan on bench_lines_sk using idx_store_day_prod over (store_sk = 1 AND 1 <= day <= 31), with index condition: ((bench_lines_sk.store_sk = 1) and (bench_lines_sk.`day` between 1 and 31))  (cost=2.21 rows=1) (actual time=0.0158..0.0158 rows=0 loops=1)


--- Composite Key (CK) with covering index ---
-> Table scan on <temporary>  (actual time=0.0434..0.0434 rows=0 loops=1)
    -> Aggregate using temporary table  (actual time=0.042..0.042 rows=0 loops=1)
        -> Filter: ((bench_li

## Covering Index Benchmark (Spend Trends)

**Note:** The `ALTER ... ADD KEY cov_store_day_prod_sales` statements were **denied**, so this run used the **existing** indexes (no new covering columns were added).

**Test:** `store_sk = 1`, `day 1..31`

### Results
| Design | Top-node time | Range-scan node time | Access Path |
|--------|---------------:|---------------------:|-------------|
| SK     | **~0.0257 s**  | ~0.0158 s            | Secondary index **`idx_store_day_prod (store, day, product)`** |
| CK     | ~0.0434 s      | ~0.0300 s            | Clustered **PRIMARY** `(store, day, product, household)` |

### Observations
- Both plans are **aligned on `(store, day)`** and are therefore extremely fast.
- In this sample, **SK is ~40% faster** overall (25.7 ms vs 43.4 ms). Its narrower secondary index scans slightly fewer bytes.
- CK uses the clustered PRIMARY range scan efficiently as well; the higher top-node time likely reflects minor temp/aggregation + cache effects.
- The plan shows **“Aggregate using temporary table”** for both, which is expected for `GROUP BY product_sk`.

### Conclusion
- For aligned predicates, **both CK and SK deliver sub-50 ms performance** with index range scans.
- **SK edged out CK** here using `idx_store_day_prod`; **CK** still performs well without needing extra secondary indexes.


## 7) Secondary index count & maintenance

In [16]:
import pymysql, pandas as pd

#  Use bench (root remote is blocked in your container)
CONN = dict(
    host="127.0.0.1",
    port=3307,
    user="bench",
    password="benchpw",
    database="retail_db"
)

def fetch_df(sql):
    with pymysql.connect(**CONN, cursorclass=pymysql.cursors.DictCursor, autocommit=True) as conn:
        with conn.cursor() as cur:
            cur.execute(sql)
            rows = cur.fetchall()
    return pd.DataFrame(rows)

# --- Sanity check (avoid confusing aliases) ---
try:
    who = fetch_df("SELECT USER() AS user_host, CURRENT_USER() AS curr_user, CONNECTION_ID() AS conn_id")
    print("Connection info:\n", who.to_string(index=False))
except Exception as e:
    print("Sanity check failed:", e)
    print(fetch_df("SELECT 1 AS ok").to_string(index=False))

# --- SHOW INDEX (raw MySQL output) ---
ck_show = fetch_df("SHOW INDEX FROM bench_lines_ck")
sk_show = fetch_df("SHOW INDEX FROM bench_lines_sk")

print("\n--- SHOW INDEX: bench_lines_ck ---")
print(ck_show.to_string(index=False))

print("\n--- SHOW INDEX: bench_lines_sk ---")
print(sk_show.to_string(index=False))

ck_show.to_csv("bench_lines_ck_show_index.csv", index=False)
sk_show.to_csv("bench_lines_sk_show_index.csv", index=False)

# --- information_schema.STATISTICS (clean detail view) ---
q_stats = """
SELECT 
  TABLE_NAME, INDEX_NAME, NON_UNIQUE, SEQ_IN_INDEX, COLUMN_NAME, SUB_PART, INDEX_TYPE, CARDINALITY
FROM information_schema.STATISTICS
WHERE TABLE_SCHEMA = DATABASE()
  AND TABLE_NAME IN ('bench_lines_ck','bench_lines_sk')
ORDER BY TABLE_NAME, INDEX_NAME, SEQ_IN_INDEX
"""
idx_details = fetch_df(q_stats)

print("\n--- information_schema.STATISTICS ---")
print(idx_details.to_string(index=False))

idx_details.to_csv("index_details.csv", index=False)
print("\n Saved: bench_lines_ck_show_index.csv, bench_lines_sk_show_index.csv, index_details.csv")


Connection info:
        user_host curr_user  conn_id
bench@172.24.0.1   bench@%     3557

--- SHOW INDEX: bench_lines_ck ---
         Table  Non_unique Key_name  Seq_in_index  Column_name Collation  Cardinality Sub_part Packed Null Index_type Comment Index_comment Visible Expression
bench_lines_ck           0  PRIMARY             1     store_sk         A          407     None   None           BTREE                           YES       None
bench_lines_ck           0  PRIMARY             2          day         A        52770     None   None           BTREE                           YES       None
bench_lines_ck           0  PRIMARY             3   product_sk         A      1410659     None   None           BTREE                           YES       None
bench_lines_ck           0  PRIMARY             4 household_sk         A      1417842     None   None           BTREE                           YES       None

--- SHOW INDEX: bench_lines_sk ---
         Table  Non_unique           Key_na

## Index Inventory (CK vs SK) — from `SHOW INDEX`

### `bench_lines_ck` (Composite Key)
| Index   | Columns (order)                                             | Type  | Cardinality (est.)                    |
|---------|--------------------------------------------------------------|-------|---------------------------------------|
| PRIMARY | `store_sk` → `day` → `product_sk` → `household_sk`           | BTREE | 407 / 52,770 / 1,410,659 / 1,417,842  |

> **Only one** clustered PRIMARY KEY; **no secondary indexes**.

---

### `bench_lines_sk` (Surrogate Key)
| Index              | Columns (order)                        | Type  | Cardinality (est.)      |
|--------------------|----------------------------------------|-------|-------------------------|
| PRIMARY            | `id`                                   | BTREE | 1,417,143               |
| `idx_store_day_prod` | `store_sk` → `day` → `product_sk`    | BTREE | 393 / 53,800 / 1,411,597|
| `idx_household`    | `household_sk`                         | BTREE | 796                     |

> **Three** indexes total: compact PK (`id`) **+ 2 secondaries** that power store/day and household access paths.

---

## Why this matters 

- **Index count & size**
  - CK: 1 index → smaller footprint (**~98.7 MB** total).
  - SK: 3 indexes → larger footprint (**~184.8 MB** total).

- **Write & maintenance overhead**
  - **Bulk insert (≈1.42M rows):** CK **23,433 ms** vs SK **37,729 ms** (SK slower due to extra index maintenance).
  - **OPTIMIZE/REBUILD:** CK **27 s** vs SK **49 s** (SK ~1.8× slower).

- **Query behavior explained**
  - **Store + day** workloads:  
    - CK uses clustered PK prefix → very fast.  
    - SK uses `idx_store_day_prod` → similarly fast (needs the secondary index).
  - **Household-centric lookups:**  
    - CK lacks an index on `household_sk` alone → scan-heavy.  
    - SK uses `idx_household` → direct index lookup (orders of magnitude faster).


## 8) Buffer pool friendliness (rough check)

In [17]:
def fetch(sql, one=False):
    with pymysql.connect(**CONN, cursorclass=pymysql.cursors.DictCursor, autocommit=True) as conn:
        with conn.cursor() as cur:
            cur.execute(sql)
            return cur.fetchone() if one else cur.fetchall()

def get_bp_stats():
    rows = fetch("SHOW GLOBAL STATUS LIKE 'Innodb_buffer_pool_read%';")
    # Rows come back as [{'Variable_name': 'Innodb_buffer_pool_read_requests', 'Value': '...'}, ...]
    d = {r["Variable_name"]: int(r["Value"]) for r in rows}
    return pd.Series(d, name="bp")

def run(sql):
    # Execute a statement and return the first row (if any)
    return fetch(sql, one=True)

# Pick a household and a store/day window
hh = fetch("SELECT household_sk FROM bench_lines_data LIMIT 1", one=True)["household_sk"]
d0 = fetch("SELECT MIN(`day`) AS d0 FROM bench_lines_data", one=True)["d0"]
d1 = d0 + 30
store = fetch("SELECT store_sk FROM bench_lines_data LIMIT 1", one=True)["store_sk"]
print(f"Using: household_sk={hh}, store_sk={store}, day range {d0}..{d1}")

# -------------------------
# 1) Buffer-pool baseline
# -------------------------
bp0 = get_bp_stats()

# -------------------------
# 2) Heavy CK query (forces scan)
# -------------------------
# household filter on CK (household_sk is the 4th PK column -> table scan)
run(f"EXPLAIN ANALYZE SELECT SUM(qty) FROM bench_lines_ck WHERE household_sk={hh}")
run(f"SELECT SUM(qty) FROM bench_lines_ck WHERE household_sk={hh}")

bp1 = get_bp_stats()

# -------------------------
# 3) Heavy SK query (uses index on household)
# -------------------------
run(f"EXPLAIN ANALYZE SELECT SUM(qty) FROM bench_lines_sk WHERE household_sk={hh}")
run(f"SELECT SUM(qty) FROM bench_lines_sk WHERE household_sk={hh}")

bp2 = get_bp_stats()

# Build comparison table
ck_diff = (bp1 - bp0).rename("after_CK_minus_before")
sk_diff = (bp2 - bp1).rename("after_SK_minus_after_CK")
cmp_df = pd.concat([bp0.rename("baseline"), bp1.rename("after_CK"), ck_diff, bp2.rename("after_SK"), sk_diff], axis=1)
print("\nInnoDB buffer-pool counters (diffs show impact of each heavy query):")
print(cmp_df.to_string())

# Save if you want to include in report
cmp_df.to_csv("buffer_pool_diffs.csv")
print("\n Saved buffer pool diffs to buffer_pool_diffs.csv")

# -------------------------
# 4) Per-table stats via sys
# -------------------------
q_sys = """
SELECT *
FROM sys.schema_table_statistics
WHERE table_schema='retail_db'
  AND table_name IN ('bench_lines_ck','bench_lines_sk')
ORDER BY rows_fetched DESC
LIMIT 10
"""
try:
    sys_df = pd.DataFrame(fetch(q_sys))
    print("\nPer-table stats (sys.schema_table_statistics):")
    print(sys_df.to_string(index=False))
    sys_df.to_csv("schema_table_statistics.csv", index=False)
    print("\n Saved schema_table_statistics.csv")
except Exception as e:
    print("\n(sys.schema_table_statistics not accessible?)", e)
    print("You can run this inside the container as root if needed:\n"
          "  SELECT * FROM sys.schema_table_statistics WHERE table_schema='retail_db' "
          "AND table_name IN ('bench_lines_ck','bench_lines_sk') ORDER BY rows_fetched DESC LIMIT 10;")

Using: household_sk=392, store_sk=1, day range 1..31

InnoDB buffer-pool counters (diffs show impact of each heavy query):
                                        baseline   after_CK  after_CK_minus_before   after_SK  after_SK_minus_after_CK
Innodb_buffer_pool_read_ahead_rnd              0          0                      0          0                        0
Innodb_buffer_pool_read_ahead             553693     553821                    128     553821                        0
Innodb_buffer_pool_read_ahead_evicted       3248       3345                     97       3345                        0
Innodb_buffer_pool_read_requests       286958218  286984248                  26030  286992357                     8109
Innodb_buffer_pool_reads                  283510     296000                  12490     296089                       89

 Saved buffer pool diffs to buffer_pool_diffs.csv

(sys.schema_table_statistics not accessible?) (1142, "SELECT command denied to user 'bench'@'172.24.0.1' for ta

## InnoDB Buffer-Pool Impact (CK vs SK — household lookup)

**Context**  
CK query = `SUM(qty) FROM bench_lines_ck WHERE household_sk=?` (no useful index → table scan)  
SK query = `SUM(qty) FROM bench_lines_sk WHERE household_sk=?` (uses `idx_household`)  
Run with: `household_sk=392`, `store_sk=1`, `day 1..31`

### Deltas in Buffer-Pool Counters
| Metric                                   | CK Δ (after_CK − baseline) | SK Δ (after_SK − after_CK) |
|------------------------------------------|----------------------------:|---------------------------:|
| Innodb_buffer_pool_read_requests         | **+26,030**                 | **+8,109**                 |
| Innodb_buffer_pool_reads *(disk misses)* | **+12,490**                 | **+89**                    |
| Innodb_buffer_pool_read_ahead            | **+128**                    | **0**                      |
| Innodb_buffer_pool_read_ahead_evicted    | **+97**                     | **0**                      |
| Innodb_buffer_pool_read_ahead_rnd        | 0                           | 0                          |

### What this shows
- The **CK** household lookup triggers a **large scan**, causing ~**3.2×** more logical requests and ~**140×** more disk reads (12,490 vs 89). Read-ahead activity confirms sequential scanning.
- The **SK** lookup uses the `idx_household` secondary index, touching far fewer pages and causing almost no misses — **much more cache-friendly**.

### Conclusion
For household-centric queries, **SK is dramatically more efficient** in buffer usage and IO due to targeted secondary indexing. CK remains space-lean, but when predicates don’t align with its leftmost PK prefix, it becomes **scan-heavy**.

> Absolute counts depend on cache warm-up; the **relative gap** (CK ≫ SK) is the key signal.


## 9) DELETE/PURGE behavior

#### Need to run the below before doing this to get DEL access -
docker exec -it sk_mysql mysql -uroot -p -e ^ </br>
"GRANT CREATE, INSERT, DELETE, DROP ON retail_db.* TO 'bench'@'%'; FLUSH PRIVILEGES;"

Verify using: </br>
docker exec -it sk_mysql mysql -uroot -p -e "SHOW GRANTS FOR 'bench'@'%';"


In [18]:
with pymysql.connect(**CONN, cursorclass=pymysql.cursors.DictCursor, autocommit=True) as conn:
    with conn.cursor() as cur:
        # 1) choose a store with lots of data (top 1 by rows)
        cur.execute("""
            SELECT store_sk
            FROM (
              SELECT store_sk, COUNT(*) AS c
              FROM bench_lines_data
              GROUP BY store_sk
              ORDER BY c DESC
              LIMIT 1
            ) t
        """)
        store = cur.fetchone()["store_sk"]

        # 2) build tiny working copies (fresh each run)
        cur.execute("DROP TABLE IF EXISTS del_ck")
        cur.execute("DROP TABLE IF EXISTS del_sk")
        cur.execute("CREATE TABLE del_ck LIKE bench_lines_ck")
        cur.execute("CREATE TABLE del_sk LIKE bench_lines_sk")
        cur.execute(f"INSERT INTO del_ck SELECT * FROM bench_lines_ck WHERE store_sk={store} LIMIT 50000")
        cur.execute(f"""
            INSERT INTO del_sk (id, store_sk, `day`, product_sk, household_sk, sales, qty)
            SELECT NULL, store_sk, `day`, product_sk, household_sk, sales, qty
            FROM bench_lines_sk
            WHERE store_sk={store}
            LIMIT 50000
        """)

        # Count rows we copied
        cur.execute("SELECT COUNT(*) AS n FROM del_ck"); n_ck_before = cur.fetchone()["n"]
        cur.execute("SELECT COUNT(*) AS n FROM del_sk"); n_sk_before = cur.fetchone()["n"]

        # 3) pick a 30-day window **from the data we just copied** (ensures some deletions happen)
        cur.execute("SELECT MIN(`day`) AS d0 FROM del_ck")
        d0 = cur.fetchone()["d0"]
        d1 = d0 + 30

        # 4) time deletes on the server and capture affected rows
        cur.execute("SET @t0 := NOW(6)")
        cur.execute(f"DELETE FROM del_ck WHERE `day` BETWEEN {d0} AND {d1}")
        cur.execute("SET @d_ck := ROW_COUNT()")
        cur.execute("SET @t1 := NOW(6)")

        cur.execute(f"DELETE FROM del_sk WHERE `day` BETWEEN {d0} AND {d1}")
        cur.execute("SET @d_sk := ROW_COUNT()")
        cur.execute("SET @t2 := NOW(6)")

        cur.execute("""
            SELECT 
              TIMESTAMPDIFF(MICROSECOND,@t0,@t1)/1000.0 AS ms_del_ck,
              @d_ck AS rows_del_ck,
              TIMESTAMPDIFF(MICROSECOND,@t1,@t2)/1000.0 AS ms_del_sk,
              @d_sk AS rows_del_sk
        """)
        res = cur.fetchone()

# Print results
print(f"Store {store}, delete window: day {d0}..{d1}")
print(pd.DataFrame([{
    "assignment_task": "delete_window",
    "design": "ck",
    "ms": round(res['ms_del_ck'], 2),
    "rowcount": int(res['rows_del_ck'])
}, {
    "assignment_task": "delete_window",
    "design": "sk",
    "ms": round(res['ms_del_sk'], 2),
    "rowcount": int(res['rows_del_sk'])
}]).to_string(index=False))

# 5) append to benchmark_results.csv
rows = [
    {"assignment_task": "delete_window", "design": "ck", "ms": round(res['ms_del_ck'], 2), "rowcount": int(res['rows_del_ck'])},
    {"assignment_task": "delete_window", "design": "sk", "ms": round(res['ms_del_sk'], 2), "rowcount": int(res['rows_del_sk'])},
]
new_df = pd.DataFrame(rows)

out_path = "delete-purge_results.csv"
if os.path.exists(out_path):
    base = pd.read_csv(out_path)
    combined = pd.concat([base, new_df], ignore_index=True)
else:
    combined = new_df
combined.to_csv(out_path, index=False)
print(f"\n Appended delete timings to {out_path}")


Store 140, delete window: day 6..36
assignment_task design     ms  rowcount
  delete_window     ck  91.00       208
  delete_window     sk 118.51       208

 Appended delete timings to benchmark_results.csv


## Delete Range Benchmark

**Setup:** Copied ~50k rows per design into `del_ck` / `del_sk` for one store, then deleted a 30-day window.  
**Window:** `store = 140`, `days 6..36`

### Results
| Design | Delete Time (ms) | Rows Deleted |
|--------|------------------:|-------------:|
| CK     | **91.00**         | 208          |
| SK     | 118.51            | 208          |

### Observations
- Both designs perform similarly for this **localized** delete (only ~208 rows).
- **CK is ~1.30× faster** (91 ms vs 118.5 ms). Rows are clustered by `(store_sk, day, …)`, so the range maps to a contiguous span.
- **SK** must also update two secondary indexes (`idx_store_day_prod`, `idx_household`), adding a small overhead.

### Takeaways
- For **range deletes aligned** with `(store, day)`, CK can be modestly faster.
- On SK, keep the predicate selective (include `store_sk`) so the optimizer can use `idx_store_day_prod`.
- For **larger deletions**, batch in chunks (e.g., `LIMIT 10k` in a loop) or consider day partitioning to reduce locks and purge work.

## 10) Table rebuild/OPTIMIZE time

In [19]:
with pymysql.connect(**CONN, cursorclass=pymysql.cursors.DictCursor, autocommit=True) as conn:
    with conn.cursor() as cur:
        cur.execute("SET @t0 := NOW(6)")
        cur.execute("OPTIMIZE TABLE bench_lines_ck")
        ck_msg = cur.fetchall()  # [{'Table':..., 'Op':'optimize', 'Msg_type':'status', 'Msg_text':'OK'|...}]

        cur.execute("SET @t1 := NOW(6)")
        cur.execute("OPTIMIZE TABLE bench_lines_sk")
        sk_msg = cur.fetchall()

        cur.execute("SET @t2 := NOW(6)")
        cur.execute("""
            SELECT 
              TIMESTAMPDIFF(SECOND,@t0,@t1) AS sec_opt_ck,
              TIMESTAMPDIFF(SECOND,@t1,@t2) AS sec_opt_sk
        """)
        times = pd.DataFrame(cur.fetchall())

print("OPTIMIZE messages (CK):")
print(pd.DataFrame(ck_msg).to_string(index=False))
print("\nOPTIMIZE messages (SK):")
print(pd.DataFrame(sk_msg).to_string(index=False))

print("\nTimings (seconds):")
print(times.to_string(index=False))
times.to_csv("optimize_times.csv", index=False)
print("\n Saved to optimize_times.csv")


OPTIMIZE messages (CK):
                   Table       Op Msg_type                                                          Msg_text
retail_db.bench_lines_ck optimize     note Table does not support optimize, doing recreate + analyze instead
retail_db.bench_lines_ck optimize   status                                                                OK

OPTIMIZE messages (SK):
                   Table       Op Msg_type                                                          Msg_text
retail_db.bench_lines_sk optimize     note Table does not support optimize, doing recreate + analyze instead
retail_db.bench_lines_sk optimize   status                                                                OK

Timings (seconds):
 sec_opt_ck  sec_opt_sk
         25          53

 Saved to optimize_times.csv


## Table Maintenance Benchmark — `OPTIMIZE TABLE`

**MySQL output:**  
> *“Table does not support optimize, doing recreate + analyze instead”*  
For InnoDB this means MySQL **rebuilds** the table and **ANALYZE**s its indexes (locks table for the duration).

### Results
| Table             | Time (s) | Notes |
|-------------------|---------:|------|
| `bench_lines_ck`  | **27**   | Fewer/lighter indexes to rebuild |
| `bench_lines_sk`  | **49**   | More/larger secondary indexes → slower |

### Observations
- The **SK** design takes ~**1.8×** longer to rebuild/analyze than **CK**.  
- This aligns with earlier size findings (SK ≈ **185 MB** vs CK ≈ **99 MB**) and SK’s extra secondary indexes; more bytes and indexes to recreate ⇒ longer maintenance.

### Takeaways
- **Maintenance overhead:** SK is heavier for housekeeping tasks (OPTIMIZE/ALTER/REBUILD) due to index fan-out and size.  
- **When to run:** after large bulk **DELETE/INSERT** waves or periodic housekeeping to defragment/reclaim per-table space.  
- **Operational note:** expect **blocking** during the rebuild; schedule during low-traffic windows or use partitioning/online DDL strategies if needed.


##  One combined Excel file

In [21]:
import os, time
import pandas as pd

# Where to save the Excel workbook
OUT_XLSX = "retail_benchmark_report.xlsx"

# Map sheet names -> CSV files you’ve been generating
CANDIDATE_SHEETS = {
    "benchmarks":        "benchmark_results.csv",           # main table with all query timings
    "insert_throughput": "insert_throughput.csv",
    "table_sizes":       "table_sizes.csv",
    "index_details":     "index_details.csv",
    "indexes_ck":        "bench_lines_ck_show_index.csv",
    "indexes_sk":        "bench_lines_sk_show_index.csv",
    "buffer_pool":       "buffer_pool_diffs.csv",
    "optimize_times":    "optimize_times.csv",
    # add more here if you create additional CSVs
}

# Collect any CSVs that actually exist
to_write = {}
for sheet, csvfile in CANDIDATE_SHEETS.items():
    if csvfile and os.path.exists(csvfile):
        to_write[sheet] = pd.read_csv(csvfile)

# Optional: a small meta sheet (only adds rows if vars exist)
meta_rows = []
if "STORE" in globals():  meta_rows.append(("STORE", STORE))
if "DAY_LO" in globals(): meta_rows.append(("DAY_LO", DAY_LO))
if "DAY_HI" in globals(): meta_rows.append(("DAY_HI", DAY_HI))
meta_rows.append(("GENERATED_AT", time.strftime("%Y-%m-%d %H:%M:%S")))
if meta_rows:
    to_write["meta"] = pd.DataFrame(meta_rows, columns=["key", "value"])

# Create the Excel file
try:
    writer = pd.ExcelWriter(OUT_XLSX, engine="xlsxwriter")
    supports_fmt = True
except Exception:
    # Fallback if xlsxwriter isn’t installed
    writer = pd.ExcelWriter(OUT_XLSX, engine="openpyxl")
    supports_fmt = False

for sheet_name, df in to_write.items():
    df.to_excel(writer, sheet_name=sheet_name, index=False)
    if supports_fmt:
        ws = writer.sheets[sheet_name]
        # freeze header row
        ws.freeze_panes(1, 0)
        # add autofilter
        if df.shape[1] > 0:
            ws.autofilter(0, 0, max(len(df), 1), df.shape[1]-1)
        # autosize columns (cap width to 60)
        for col_idx, col in enumerate(df.columns):
            maxlen = max([len(str(col))] + [len(str(v)) for v in df[col].astype(str).tolist()]) if len(df) else len(str(col))
            ws.set_column(col_idx, col_idx, min(maxlen + 2, 60))

writer.close()
print(f" Wrote {OUT_XLSX} with sheets: {', '.join(to_write.keys())}")


 Wrote retail_benchmark_report.xlsx with sheets: benchmarks, table_sizes, index_details, indexes_ck, indexes_sk, buffer_pool, optimize_times, meta


# Executive Summary — 4 Deliverables (Store 140, Days 6..36)

| Task              | CK (ms) | SK (ms) | Rows | Faster |
|-------------------|--------:|--------:|-----:|:------:|
| spend_trends      |   5.47  |   6.63  |  177 | **CK** |
| demo_influence    | 563.75  | 488.05  |   10 | **SK** |
| growth_categories | 1561.09 | 1247.66 |   10 | **SK** |
| campaign_impact   | 6744.19 | 6119.41 |    2 | **SK** |

**Summary:** SK wins **3/4** deliverables (joins/aggregations); CK wins **spend_trends** because the query is perfectly aligned with CK’s clustered key `(store, day, product, …)`.



## Deliverables — What was measured, why, and why each result makes sense

### 1) Spend trends — revenue by product (winner: **CK**, 5.47 ms vs 6.63 ms; 177 rows)
- **Query shape:** `WHERE store_sk=? AND day BETWEEN ? AND ?  GROUP BY product_sk`
- **Why include:** Bread-and-butter retail slice (recent window per store).
- **Access path:**  
  - **CK:** clustered PRIMARY KEY `(store, day, product, …)` → **perfect left-prefix match** → tight range scan; row already on leaf → minimal lookups.  
  - **SK:** secondary `idx_store_day_prod(store, day, product)` → also fast, but one more structure to maintain.
- **Why CK won:** Storage order matches the predicate; reads are sequential and naturally “covering” (clustered index holds the row).

**If asked “How to make SK equal/better?”** Add/keep `idx_store_day_prod`; consider including `sales` to make it covering (trade: more index bytes).

---

### 2) Demographic influence — revenue by income × age (winner: **SK**, 488.05 ms vs 563.75 ms; 10 rows)
- **Query shape:** scan fact → join `household_dim` by `household_sk` → `GROUP BY income_desc, age_desc` → `ORDER BY revenue DESC LIMIT 10`.
- **Why include:** Classic **join + aggregate** used by marketing.
- **Access path:** Both designs **scan ~1.42M rows** and do **cheap PK lookups** into the dimension, then aggregate & sort.
- **Why SK edged out:** Mostly scan/aggregate dominated; SK’s row layout and narrower secondary index can shave a bit of work. Gap is modest and often sensitive to buffer-pool warm-up.

**Tuning levers:** Pre-aggregate by household period; add covering pieces if repeatedly selecting the same columns.

---

### 3) Growth categories — 30-day vs prior 28-day (winner: **SK**, 1,247.66 ms vs 1,561.09 ms; 10 rows)
- **Query shape:** two windowed scans on fact → join `product_dim` → compare `SUM(sales)` by category; `ORDER BY` growth.
- **Why include:** Common PoP growth analysis.
- **Access path:** Two large scans + join + group.  
- **Why SK edged out:** Similar reason as #2—slightly less I/O via compact secondary index paths; still largely scan/aggregate bound.

**Tuning levers:** Materialize daily category revenue; consider summary tables or column pruning/covering indexes.

---

### 4) Campaign impact — coupon redeemed vs not (winner: **SK**, 6,119.41 ms vs 6,744.19 ms; 2 rows)
- **Query shape:** match redemption to fact on `(household, product, day)` → split **redeemed** vs **not_redeemed** → aggregate revenue/units.
- **Why include:** Typical marketing effectiveness readout.
- **Access path:** Big join + two aggregates (one anti-join).  
- **Why SK edged out:** Extra flexibility from secondary indexes; but workload is **data-volume dominated**, so the delta is moderate.

**Tuning levers:** Precompute redemption flags per (store, day, product); use temp tables or summaries for frequent reporting windows.

---

### Overall
- “**SK wins 3/4 deliverables** because those are **join/aggregate** heavy and benefit from flexible secondary indexes; **CK wins spend_trends** because that query is **perfectly aligned** to its clustered order `(store → day → product → …)`.”



# 10 Benchmark Scenarios (Notebook “parameters”)

| # | Scenario / What we measured | Query/Feature exercised | Metric(s) captured | Winner (this run) |
|---|-----------------------------|-------------------------|--------------------|-------------------|
| 1 | **Spend trends** (store + 30-day window → revenue by product) | Range scan on `(store, day)` + `GROUP BY product` | ms, rowcount | **CK** (5.47 vs 6.63 ms) |
| 2 | **Demographic influence** (income × age → revenue) | Scan fact + join `household_dim` + aggregate | ms, rowcount | **SK** (488 vs 564 ms) |
| 3 | **Growth categories** (30-day vs prior 28-day) | Two windowed scans + join `product_dim` | ms, rowcount | **SK** (1248 vs 1561 ms) |
| 4 | **Campaign impact** (coupon redeemed vs not) | Join fact ↔ redemption + aggregate | ms, rowcount | **SK** (6119 vs 6744 ms) |
| 5 | **Household lookup** (`SUM(qty)` by `household_sk`) | CK: no index → scan; SK: `idx_household` | ms, rows examined | **SK** (~7.8 ms vs ~1444 ms) |
| 6 | **Store + day micro** (same window as #1) | Pure index range scan timing | ms (top & range node) | **SK** (~26 ms vs ~43 ms) |
| 7 | **Insert throughput** (~1.42M rows) | Bulk insert into each design | ms, rows inserted | **CK** (23,434 vs 37,729 ms) |
| 8 | **Delete window** (store 140, days 6..36) | Localized range delete | ms, rows deleted | **CK** (91 vs 119 ms) |
| 9 | **Storage footprint** | `information_schema.tables` | data_mb, index_mb, total_mb | **CK** (98.7 MB vs 184.8 MB) |
|10 | **Maintenance / OPTIMIZE** | Table rebuild + analyze | seconds | **CK** (25 s vs 53 s) |

*(Bonus check we also ran: buffer-pool impact — CK created ~26k extra read requests and ~12,490 disk reads vs SK’s ~8k and ~89 for the household lookup; SK is far more cache-friendly for that access pattern.)*